core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)

Building ldc fails on riscv64:

core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs0, 0(a2)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs1, 0(a2)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs2, 0(a2)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs3, 0(a2)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs4, 0(a2)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs5, 0(a2)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs6, 0(a2)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs7, 0(a2)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs8, 0(a2)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs9, 0(a2)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs10, 0(a2)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs11, 0(a2)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs0, 0(a0)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs1, 0(a0)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs2, 0(a0)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs3, 0(a0)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs4, 0(a0)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs5, 0(a0)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs6, 0(a0)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs7, 0(a0)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs8, 0(a0)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs9, 0(a0)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs10, 0(a0)
        ^
core/thread/osthread.d(1594):1:2: error: instruction requires the following: 'D' (Double-Precision Floating-Point)
        fsd fs11, 0(a0)
        ^
make[2]: *** [runtime/CMakeFiles/druntime-ldc-shared.dir/build.make:346: runtime/objects-shared/core/atomic.o] Error 1
make[2]: Leaving directory '/home/abuild/rpmbuild/BUILD/ldc-1.32.0-src/build-bootstrap'
make[1]: *** [CMakeFiles/Makefile2:493: runtime/CMakeFiles/druntime-ldc-shared.dir/all] Error 2
make: *** [Makefile:169: all] Error 2

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@
 
 #### Platform support
 - Supports LLVM 11.0 - 16.0.
+- When nothing else specified now defaults to `-mattr=+m,+a,+f,+d,+c` when targeting hosted 64-bit RISC-V. (#4390)
 
 #### Bug fixes
 - Fix function pointers/delegates on Harvard architectures (e.g., AVR). (#4432, #4465)

--- a/driver/targetmachine.cpp
+++ b/driver/targetmachine.cpp
@@ -57,7 +57,7 @@ static llvm::cl::opt<bool, true> preserveDwarfLineSection(
     llvm::cl::init(false));
 #endif
 
-const char *getABI(const llvm::Triple &triple) {
+const char *getABI(const llvm::Triple &triple, const llvm::SmallVector<llvm::StringRef, 8> &features) {
   llvm::StringRef ABIName(opts::mABI);
   if (ABIName != "") {
     switch (triple.getArch()) {
@@ -111,6 +111,17 @@ const char *getABI(const llvm::Triple &triple) {
             ABIName.str().c_str());
   }
 
+  // checks if the features include ±<feature>
+  auto hasFeature = [&features](llvm::StringRef feature) {
+    for (int i = features.size() - 1; i >= 0; i--) {
+      auto f = features[i];
+      if (f.substr(1) == feature) {
+	return f[0] == '+';
+      }
+    }
+    return false;
+  };
+
   switch (triple.getArch()) {
   case llvm::Triple::mips64:
   case llvm::Triple::mips64el:
@@ -120,6 +131,10 @@ const char *getABI(const llvm::Triple &triple) {
   case llvm::Triple::ppc64le:
     return "elfv2";
   case llvm::Triple::riscv64:
+    if (hasFeature("d"))
+      return "lp64d";
+    if (hasFeature("f"))
+      return "lp64f";
     return "lp64";
   case llvm::Triple::riscv32:
     return "ilp32";
@@ -430,15 +445,31 @@ createTargetMachine(const std::string targetTriple, const std::string arch,
 
   // checks if the features include ±<feature>
   auto hasFeature = [&features](llvm::StringRef feature) {
-    return std::any_of(
-        features.begin(), features.end(),
-        [feature](llvm::StringRef f) { return f.substr(1) == feature; });
+    for (int i = features.size() - 1; i >= 0; i--) {
+      auto f = features[i];
+      if (f.substr(1) == feature) {
+	return f[0] == '+';
+      }
+    }
+    return false;
   };
 
   // cmpxchg16b is not available on old 64bit CPUs. Enable code generation
   // if the user did not make an explicit choice.
   if (cpu == "x86-64" && !hasFeature("cx16")) {
     features.push_back("+cx16");
+  }
+
+  // For a hosted RISC-V 64-bit target default to rv64gc if nothing has
+  // been selected
+  if (triple.getArch() == llvm::Triple::riscv64 &&
+      triple.getOS() != llvm::Triple::UnknownOS &&
+      features.empty()) {
+    features.push_back("+m");
+    features.push_back("+a");
+    features.push_back("+f");
+    features.push_back("+d");
+    features.push_back("+c");
   }
 
   // Handle cases where LLVM picks wrong default relocModel
@@ -478,7 +509,7 @@ createTargetMachine(const std::string targetTriple, const std::string arch,
       opts::InitTargetOptionsFromCodeGenFlags(triple);
 
   if (targetOptions.MCOptions.ABIName.empty())
-    targetOptions.MCOptions.ABIName = getABI(triple);
+    targetOptions.MCOptions.ABIName = getABI(triple, features);
 
   if (floatABI == FloatABI::Default) {
     switch (triple.getArch()) {

--- a/driver/targetmachine.h
+++ b/driver/targetmachine.h
@@ -15,6 +15,8 @@
 #pragma once
 
 #include "llvm/ADT/Optional.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
 #include "llvm/Support/CodeGen.h"
 #include <string>
 #include <vector>
@@ -73,4 +75,5 @@ MipsABI::Type getMipsABI();
 const llvm::Target *lookupTarget(const std::string &arch, llvm::Triple &triple,
                                  std::string &errorMsg);
 
-const char *getABI(const llvm::Triple &triple);
+const char *getABI(const llvm::Triple &triple,
+		   const llvm::SmallVector<llvm::StringRef, 8> &features);

--- a/driver/tool.cpp
+++ b/driver/tool.cpp
@@ -123,13 +123,13 @@ void appendTargetArgsForGcc(std::vector<std::string> &args) {
 
   case Triple::riscv64:
     {
-      std::string mabi = getABI(triple);
-      args.push_back("-mabi=" + mabi);
-
       extern llvm::TargetMachine* gTargetMachine;
       auto featuresStr = gTargetMachine->getTargetFeatureString();
       llvm::SmallVector<llvm::StringRef, 8> features;
       featuresStr.split(features, ",", -1, false);
+
+      std::string mabi = getABI(triple, features);
+      args.push_back("-mabi=" + mabi);
 
       // Returns true if 'feature' is enabled and false otherwise. Handles the
       // case where the feature is specified multiple times ('+m,-m'), and

--- a/tests/driver/riscv_abi4.d
+++ b/tests/driver/riscv_abi4.d
@@ -1,0 +1,6 @@
+// REQUIRES: target_RISCV
+
+// RUN: %ldc %s -mtriple=riscv64-unknown-linux --gcc=echo > %t && FileCheck %s < %t
+// CHECK: -mabi=lp64d -march=rv64gc
+
+void main() {}

--- a/tests/driver/riscv_abi5.d
+++ b/tests/driver/riscv_abi5.d
@@ -1,0 +1,6 @@
+// REQUIRES: target_RISCV
+
+// RUN: %ldc %s -mtriple=riscv64-unknown-linux -mattr=+m,+a,+f,-d --gcc=echo > %t && FileCheck %s < %t
+// CHECK: -mabi=lp64f -march=rv64imaf_zicsr_zifencei
+
+void main() {}

